src/baldr/curler.cc

```diff
@@ -0,0 +1,123 @@
+  if (!out) {
+    return static_cast<size_t>(0);
+  }
+  out->insert(out->end(), in, in + (block_size * blocks));
+  return block_size * blocks;
+}
+
+} // namespace
+
+namespace valhalla {
+namespace baldr {
+
+struct curler_t::pimpl_t {
+  pimpl_t() : connection(init_curl()) {
+    if (connection.get() == nullptr) {
+      LOG_ERROR("Failed to created CURL connection");
+      throw std::runtime_error("Failed to created CURL connection");
+    }
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_ERRORBUFFER, error),
+                "Failed to set error buffer ");
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_FOLLOWLOCATION, 1L),
+                "Failed to set redirect option ");
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_WRITEFUNCTION, write_callback),
+                "Failed to set writer ");
+    // this is less secure but we'll worry about that later
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_SSL_VERIFYPEER, 0L),
+                "Failed to disable peer verification ");
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_SSL_VERIFYHOST, 0L),
+                "Failed to disable host verification ");
+  }
+  // TODO: retries?
+  std::vector<char> fetch(const std::string& url, long& http_code, bool allow_compression) {
+    // sending content encoding as "" makes curl automatically handle identity, gzip or deflate
+    // sending nullptr tells curl to only accept identity (no decoding needed)
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_ACCEPT_ENCODING,
+                                 allow_compression ? ALL_ENCODINGS : nullptr),
+                "Failed to set content encoding header ");
+    // set the url
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_URL, url.c_str()), "Failed to set URL ");
+    // set the location of the result
+    std::vector<char> result;
+    assert_curl(curl_easy_setopt(connection.get(), CURLOPT_WRITEDATA, &result),
+                "Failed to set write data ");
+    // get the url
+    assert_curl(curl_easy_perform(connection.get()), "Failed to get URL ");
+    // grab the return code
+    curl_easy_getinfo(connection.get(), CURLINFO_RESPONSE_CODE, &http_code);
+    // hand over the results
+    return result;
+  }
+  void assert_curl(CURLcode code, const std::string& msg) {
+    if (code != CURLE_OK) {
+      std::string what = msg + error;
+      LOG_ERROR(what);
+      throw std::runtime_error(what);
+    }
+  }
+  std::shared_ptr<CURL> connection;
+  char error[CURL_ERROR_SIZE];
+};
+
+curler_t::curler_t() : pimpl(new pimpl_t()) {
+}
+
+std::vector<char> curler_t::
+operator()(const std::string& url, long& http_code, bool allow_compression) {
+  return pimpl->fetch(url, http_code, allow_compression);
+}
+
+} // namespace baldr
+} // namespace valhalla
+
+#else
+
+// if you dont build with CURL support we always error when you try to use it
```

Comment:
so if you dont build with curl support we still provide the interface but when you try to use it we throw an exception. we could also log error here if need be!